File: content/courses/fortran_introduction/solns/day_of_week.f90
program day_of_week
!*******************************************************************************
  ! This program computes the day of the week given a date in the 
  ! Gregorian calendar.
  !
  ! Author:    K. Holcomb
  ! Changelog: 2013-06-05 Initial version
  !            2014-01-29 Bug fix to correct leap_year test for century 
  !            2014-01-29 Added a loop for user input
  !            2015-01-29 Modification to use only conditionals (no arrays)
!*******************************************************************************
   implicit none

   integer                          ::  max_year, min_year
   integer                          ::  day, month, year, century
   integer                          ::  D, M, Y, C, L, W
   logical                          ::  leap_year, century_leap_year


   day=2
   month=3
   year=2016

   D=day
   century=100*(int(year)/100)
   Y=year-century

   century_leap_year = mod(century,400)==0

   leap_year=.false.
   if (Y>0) then
      leap_year = mod(year,4)==0
   else
      if (century_leap_year) leap_year=.true.
   endif

   L = int(Y)/4

   if (century_leap_year) L=L+1

   if (leap_year .and. month<3) L=L-1

   if (month==1 .or. month==10) then
      M=0
   elseif (month==2 .or. month==3 .or. month==11) then
      M=3
   elseif (month==4 .or. month==7) then
      M=6
   elseif (month==5) then
      M=1
   elseif (month==6) then
      M=4
   elseif (month==8) then
      M=2
   else
      M=5
   endif

   if     ( century==1400 .or. century==1800 .or. century==2200 ) then
      C=2
   elseif ( century==1500 .or. century==1900 .or. century==2300 ) then
      C=0
   elseif ( century==1600 .or. century==2000 .or. century==2400 ) then
      C=5
   elseif ( century==1700 .or. century==2100 .or. century==2500 ) then
      C=4
   else
      print *, "This algorithm doesn't cover the century requested"
   endif

   W=mod((C+Y+L+M+D),7)

   !Example of select case
   select case(W)
   case(0)
      print *, 'Sunday'
   case(1)
      print *, 'Monday'
   case(2)
      print *, 'Tuesday'
   case(3)
      print *, 'Wednesday'
   case(4)
      print *, 'Thursday'
   case(5)
      print *, 'Friday'
   case(6)
      print *, 'Saturday'
   end select 

end program

File: content/courses/cpp_introduction/solns/day_of_week.cxx
#include <iostream>

using namespace std;

/*******************************************************************************
  ! This program computes the day of the week given a date in the
  ! Gregorian calendar.
  !
  ! Author:    K. Holcomb
  ! Changelog: 2013-06-05 Initial version
  !            2014-01-29 Bug fix to correct leap_year test for century
  !            2014-01-29 Added a loop for user input
  !            2015-01-29 Modification to use only conditionals (no arrays)
*******************************************************************************/

int main() {

    int  maxYear, minYear;
    int  day, month, year, century;
    int  D, M, Y, C, L, W;
    bool leapYear, centuryLeapYear;

    day=2;
    month=3;
    year=2016;

    D=day;
    century=100*(int(year)/100);
    Y=year-century;

    centuryLeapYear=(century%400)==0;

    leapYear=false;

    if (Y>0) {
        leapYear=(year%4)==0;
    }
    else if (centuryLeapYear) {
        leapYear=true;
    }

    L=int(Y)/4;

    if (centuryLeapYear) L+=1;

    if (leapYear && month<3) L-=1;

    if (month==1 || month==10) {
        M=0;
    }
    else if (month==2 || month==3 || month==11) {
        M=3;
    }
    else if (month==4 || month==7) {
        M=6;
    }
    else if (month==5) {
        M=1;
    }
    else if (month==6) {
        M=4;
    }
    else if (month==8) {
        M=2;
    }
    else {
        M=5;
    }

    if  ( century==1400 || century==1800 || century==2200) {
        C=2;
    }
    else if  ( century==1500 || century==1900 || century==2300) {
        C=0;
    }
    else if  ( century==1600 || century==2000 || century==2400) {
        C=5;
    }
    else if  ( century==1700 || century==2100 || century==2500) {
        C=4;
    }
    else {
        cout<<"This algorithm doesn't cover the century requested.\n";
    }

    W=(C+Y+L+M+D)%7;

    switch (W) {
        case(0):
            cout<<"Sunday\n";
            break;
        case(1):
            cout<<"Monday\n";
            break;
        case(2):
            cout<<"Tuesday\n";
            break;
        case(3):
            cout<<"Wednesday\n";
            break;
        case(4):
            cout<<"Thursday\n";
            break;
        case(5):
            cout<<"Friday\n";
            break;
        case(6):
            cout<<"Saturday\n";
            break;
        case(7):
            cout<<"Sunday\n";
            break;
    }

    return 0;

}

